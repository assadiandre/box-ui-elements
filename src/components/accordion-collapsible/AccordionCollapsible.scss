@import '../../styles/variables';

.accordion-collapsible-container {
    width: 100%;
    background-color: $white;
    box-shadow: 0 0 0 1px $bdl-gray-10;

    @include bdl-transition($bdl-transitionDurationBase, background-color);

    .accordion-collapsible-header {
        height: 58px;
    }

    .accordion-collapsible-dropdown-container {
        padding-right: 15px;
        padding-bottom: 20px;
        padding-left: 25px;
        background-color: $bdl-gray-02;
    }

    .accordion-collapsible-header-btn {
        display: flex;
        align-items: center;
        justify-content: space-between;
        width: 100%;
        height: 100%;
        padding-right: 15px;
        padding-left: 25px;

        &:hover {
            padding-right: 15px;
            padding-left: 25px;
        }
    }

    .accordion-collapsible-title {
        font-weight: bold;
        font-size: 15px;
    }

    .accordion-collapsible-icon {
        @include bdl-transition($bdl-transitionDurationBase, transform);

        transform: rotateZ(90deg);
    }

    &.is-open {
        background-color: $bdl-gray-02;

        .accordion-collapsible-icon {
            transform: rotateZ(360deg);
        }
    }
}
