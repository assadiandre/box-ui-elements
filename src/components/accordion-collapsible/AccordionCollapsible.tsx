import * as React from 'react';
import classNames from 'classnames';
import AnimateHeight from 'react-animate-height';

import { bdlGray50 } from '../../styles/variables';
import IconChevron, { DirectionType } from '../../icons/general/IconChevron';
import PlainButton from '../plain-button';
import { ButtonType } from '../button';

import './AccordionCollapsible.scss';

type AccordionCollapsibleProps = {
    /** animationDuration - duration of animation in milliseconds */
    animationDuration?: number;
    /** containerProps - other props (e.g. resin target names) to be included in wrapper div */
    containerProps?: Record<string, React.ReactNode>;
    /** buttonProps - other props (e.g. resin target names) to be included in the button */
    buttonProps?: Record<string, React.ReactNode>;
    /** children - content to be displayed in the card when it is expanded */
    children: Array<React.ReactChild> | React.ReactChild;
    /** className - CSS class for the wrapper div */
    className?: string;
    /** isOpen - initial state of the collapsible card */
    isOpen: boolean;
    /** onClose - callback called when collapsible is opened */
    onClose?: Function;
    /** onOpen - callback called when collapsible is collapsed */
    onOpen?: Function;
    /** title - string or component in the title of the collapsible card */
    title: string | React.ReactElement;
};

type AccordionCollapsibleState = {
    /** isOpen - initial state of the collapsible card */
    isOpen: boolean;
};

const AccordionCollapsible = ({
    animationDuration = 100,
    containerProps,
    buttonProps,
    children,
    className,
    isOpen,
    onClose,
    onOpen,
    title,
}: AccordionCollapsibleProps) => {
    const [state, setState] = React.useState<AccordionCollapsibleState>({ isOpen });

    const toggleVisibility = () => {
        if (state.isOpen && onClose) {
            onClose();
        } else if (!state.isOpen && onOpen) {
            onOpen();
        }
        setState({ isOpen: !state.isOpen });
    };

    return (
        <div
            {...containerProps}
            className={classNames('accordion-collapsible-container', className, {
                'is-open': state.isOpen,
            })}
        >
            <div className="accordion-collapsible-header">
                <PlainButton
                    {...buttonProps}
                    aria-expanded={state.isOpen}
                    className="accordion-collapsible-header-btn"
                    onClick={toggleVisibility}
                    type={ButtonType.BUTTON}
                >
                    <div className="accordion-collapsible-title">{title}</div>
                    <IconChevron
                        className="accordion-collapsible-icon"
                        color={bdlGray50}
                        direction={state.isOpen ? DirectionType.DOWN : DirectionType.RIGHT}
                    />
                </PlainButton>
            </div>
            <AnimateHeight duration={animationDuration} height={state.isOpen ? 'auto' : 0}>
                <div className="accordion-collapsible-dropdown-container">{children}</div>
            </AnimateHeight>
        </div>
    );
};

export { AccordionCollapsibleProps };
export default AccordionCollapsible;
