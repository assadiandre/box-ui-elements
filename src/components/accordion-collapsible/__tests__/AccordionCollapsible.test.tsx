import React from 'react';
import sinon from 'sinon';
import { shallow, ShallowWrapper } from 'enzyme';
import AccordionCollapsible from '../AccordionCollapsible';

describe('components/accordion-collapsible/AccordionCollapsible', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = shallow(
            <AccordionCollapsible isOpen title="foo">
                <span className="test-content">foobar</span>
            </AccordionCollapsible>,
        );
    });

    test('should render component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should toggle visibility on click', async () => {
        wrapper.find('.accordion-collapsible-header-btn').simulate('click');
        expect(wrapper).toMatchSnapshot();

        wrapper.find('.accordion-collapsible-header-btn').simulate('click');
        expect(wrapper).toMatchSnapshot();
    });

    test('should apply correct isOpen state', () => {
        wrapper = shallow(
            <AccordionCollapsible isOpen={false} title="foo">
                <span>foobar</span>
            </AccordionCollapsible>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should apply buttonProps correctly', () => {
        wrapper = shallow(
            <AccordionCollapsible buttonProps={{ a: 1, b: 2 }} isOpen title="foo">
                <span>foobar</span>
            </AccordionCollapsible>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should apply containerProps correctly', () => {
        wrapper = shallow(
            <AccordionCollapsible containerProps={{ a: 1, b: 2 }} isOpen title="foo">
                <span>foobar</span>
            </AccordionCollapsible>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should fire open and close handlers', () => {
        const closeSpy = sinon.spy();
        const openSpy = sinon.spy();

        wrapper = shallow(
            <AccordionCollapsible buttonProps={{ a: 1, b: 2 }} isOpen onClose={closeSpy} onOpen={openSpy} title="foo">
                <span>foobar</span>
            </AccordionCollapsible>,
        );

        expect(closeSpy.notCalled).toBe(true);
        expect(openSpy.notCalled).toBe(true);

        wrapper.find('.accordion-collapsible-header-btn').simulate('click');

        expect(closeSpy.calledOnce).toBe(true);
        expect(openSpy.notCalled).toBe(true);

        wrapper.find('.accordion-collapsible-header-btn').simulate('click');
        expect(closeSpy.calledOnce).toBe(true);
        expect(openSpy.calledOnce).toBe(true);
    });
});
