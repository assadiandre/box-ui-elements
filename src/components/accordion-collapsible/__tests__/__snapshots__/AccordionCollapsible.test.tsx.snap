// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/accordion-collapsible/AccordionCollapsible should apply buttonProps correctly 1`] = `
<div
  className="accordion-collapsible-container is-open"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      a={1}
      aria-expanded={true}
      b={2}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="down"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height="auto"
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span>
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;

exports[`components/accordion-collapsible/AccordionCollapsible should apply containerProps correctly 1`] = `
<div
  a={1}
  b={2}
  className="accordion-collapsible-container is-open"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      aria-expanded={true}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="down"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height="auto"
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span>
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;

exports[`components/accordion-collapsible/AccordionCollapsible should apply correct isOpen state 1`] = `
<div
  className="accordion-collapsible-container"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      aria-expanded={false}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="right"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height={0}
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span>
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;

exports[`components/accordion-collapsible/AccordionCollapsible should render component correctly 1`] = `
<div
  className="accordion-collapsible-container is-open"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      aria-expanded={true}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="down"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height="auto"
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span
        className="test-content"
      >
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;

exports[`components/accordion-collapsible/AccordionCollapsible should toggle visibility on click 1`] = `
<div
  className="accordion-collapsible-container"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      aria-expanded={false}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="right"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height={0}
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span
        className="test-content"
      >
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;

exports[`components/accordion-collapsible/AccordionCollapsible should toggle visibility on click 2`] = `
<div
  className="accordion-collapsible-container is-open"
>
  <div
    className="accordion-collapsible-header"
  >
    <PlainButton
      aria-expanded={true}
      className="accordion-collapsible-header-btn"
      onClick={[Function]}
      type="button"
    >
      <div
        className="accordion-collapsible-title"
      >
        foo
      </div>
      <IconChevron
        className="accordion-collapsible-icon"
        color="#909090"
        direction="down"
      />
    </PlainButton>
  </div>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease"
    height="auto"
    style={Object {}}
  >
    <div
      className="accordion-collapsible-dropdown-container"
    >
      <span
        className="test-content"
      >
        foobar
      </span>
    </div>
  </AnimateHeight>
</div>
`;
